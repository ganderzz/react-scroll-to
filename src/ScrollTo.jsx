import React, { Component, createContext } from "react";

export const ScrollToContext = createContext("scrollToContext");

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window or ScrollArea component
 */
class ScrollTo extends Component {
  constructor(props) {
    super(props);

    this.scrollArea = {};

    this.getContext = {
      addScrollArea: (id, ref) => {
        this.scrollArea[id] = ref;
      },
      removeScrollArea: id => {
        delete this.scrollArea[id];
      }
    };
  }

  handleScroll = (props = {}) => {
    const scrollAreaKeys = Object.keys(this.scrollArea);
    const { id, ref, ...rest } = props;

    if (ref) {
      const refNode = ref.current ? ref.current : ref;

      // Scroll by ref
      this._scrollNode(refNode, rest);
    } else if (id) {
      // Scroll by id
      const node = this.scrollArea[id];

      this._scrollNode(node, rest);
    } else if (scrollAreaKeys.length > 0) {
      // Scroll by all scroll areas
      scrollAreaKeys.forEach(key => {
        const node = this.scrollArea[key];

        this._scrollNode(node, rest);
      });
    } else {
      // Scroll by window
      this._scrollNode(window, rest);
    }
  };

  _scrollNode = (node, options) => {
    if (!node) {
      return;
    }

    const top = options.y;
    const left = options.x;

    if (node.scrollTo) {
      node.scrollTo({
        top,
        left,
        behavior: options.smooth ? "smooth" : "auto"
      });
    } else {
      node.scrollLeft = left;
      node.scrollTop = top;
    }
  };

  render() {
    return (
      <ScrollToContext.Provider value={this.getContext}>
        {this.props.children &&
          this.props.children({
            scrollTo: this.handleScroll
          })}
      </ScrollToContext.Provider>
    );
  }
}

ScrollTo.defaultProps = {
  children: () => {}
};

export default ScrollTo;
