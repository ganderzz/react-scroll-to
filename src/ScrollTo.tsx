import * as React from "react";
import { IScrollOptions } from "./IScrollOptions";
import { scrollNode } from "./logic";
import { useScrollTo } from "./UseScrollTo";

interface IProps {
  children?: (props: {
    scroll: (props?: IScrollOptions) => void;
  }) => JSX.Element;
}

/**
 * Component that uses render props to inject
 * a function that allows the consumer to scroll to a
 * position in the window or ScrollArea component
 */
export function ScrollTo({ children }: IProps) {
  const { scroll } = useScrollTo();

  if (!children) {
    return null;
  }

  return children({ scroll });
}
