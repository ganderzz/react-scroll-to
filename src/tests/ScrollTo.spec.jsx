import React from "react";
import { render, cleanup, fireEvent } from "react-testing-library";
import { shallow } from "enzyme";
import ScrollTo from "../ScrollTo";

afterEach(cleanup);

beforeEach(() => {
  window.scrollTo = jest.fn();
});

describe("Test render prop.", () => {
  it("Should render the functional children.", () => {
    const { container } = render(<ScrollTo>{() => <div>test</div>}</ScrollTo>);

    expect(container).toMatchSnapshot();
  });

  it("Should render nothing on undefined.", () => {
    const { container } = render(<ScrollTo>{undefined}</ScrollTo>);

    expect(container).toMatchSnapshot();
  });

  it("Should call window.scrollTo.", () => {
    const { container } = render(
      <ScrollTo>
        {({ scrollTo }) => (
          <button onClick={() => scrollTo({ x: 100, y: 200 })}>test</button>
        )}
      </ScrollTo>
    );

    fireEvent.click(container.querySelector("button"));

    expect(window.scrollTo).toHaveBeenCalledTimes(1);
    expect(window.scrollTo.mock.calls[0]).toEqual([
      {
        left: 100,
        top: 200,
        behavior: "auto"
      }
    ]);
  });

  it("Should call window.scrollTo with default x,y when no arguments are provided.", () => {
    const { container } = render(
      <ScrollTo>
        {({ scrollTo }) => <button onClick={() => scrollTo()}>test</button>}
      </ScrollTo>
    );

    fireEvent.click(container.querySelector("button"));

    expect(window.scrollTo).toHaveBeenCalledTimes(1);
    expect(window.scrollTo.mock.calls[0]).toEqual([
      {
        left: undefined,
        top: undefined,
        behavior: "auto"
      }
    ]);
  });

  it("Should remove scroll area.", () => {
    const wrapper = shallow(<ScrollTo>{() => <div>Test</div>}</ScrollTo>);
    const childContext = wrapper.instance().getContext;
    childContext.addScrollArea("id", "foo");

    childContext.removeScrollArea("id");

    expect(wrapper.instance().scrollArea).toEqual({});
  });

  it("Should update scroll position of ScrollArea's if present, rather than window", () => {
    const mockNode = {
      scrollLeft: 0,
      scrollTop: 0
    };
    const wrapper = shallow(
      <ScrollTo>
        {({ scrollTo }) => (
          <button onClick={() => scrollTo({ x: 100, y: 200 })}>test</button>
        )}
      </ScrollTo>
    );
    const childContext = wrapper.instance().getContext;
    childContext.addScrollArea("id", mockNode);

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(mockNode).toMatchSnapshot();
  });

  it("Should scroll by ID", () => {
    const mockNode = {
      scrollLeft: 0,
      scrollTop: 0,
      id: "foo"
    };
    const wrapper = shallow(
      <ScrollTo>
        {({ scrollTo }) => (
          <button
            type="button"
            onClick={() => scrollTo({ id: "foo", x: 100, y: 200 })}
          >
            test
          </button>
        )}
      </ScrollTo>
    );
    const childContext = wrapper.instance().getContext;
    childContext.addScrollArea("foo", mockNode);

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(mockNode).toMatchSnapshot();
  });

  it("Should not break if scrolling by an unknown ID", () => {
    const mockNode = {
      scrollLeft: 0,
      scrollTop: 0,
      id: "foo"
    };
    const wrapper = shallow(
      <ScrollTo>
        {({ scrollTo }) => (
          <button
            type="button"
            onClick={() => scrollTo({ id: "unknown-id", x: 100, y: 200 })}
          >
            test
          </button>
        )}
      </ScrollTo>
    );
    const childContext = wrapper.instance().getContext;
    childContext.addScrollArea("foo", mockNode);

    const buttonEl = wrapper.find("button");
    buttonEl.simulate("click");

    expect(mockNode).toMatchSnapshot();
  });

  it("Should use smooth scrolling when enabled", () => {
    const { getByText } = render(
      <ScrollTo>
        {({ scrollTo }) => (
          <button
            type="button"
            onClick={() => scrollTo({ x: 100, y: 200, smooth: true })}
          >
            myBtn
          </button>
        )}
      </ScrollTo>
    );

    const buttonEl = getByText("myBtn");
    fireEvent.click(buttonEl);

    expect(window.scrollTo).toHaveBeenCalledTimes(1);
    expect(window.scrollTo.mock.calls[0]).toEqual([
      {
        left: 100,
        top: 200,
        behavior: "smooth"
      }
    ]);
  });

  it("Should scroll by ref when provided", () => {
    const refDOM = React.createRef();
    const { getByText } = render(
      <ScrollTo>
        {({ scrollTo }) => (
          <>
            <button
              type="button"
              onClick={() => scrollTo({ ref: refDOM, x: 100, y: 200 })}
            >
              myBtn
            </button>

            <div ref={refDOM}>test</div>
          </>
        )}
      </ScrollTo>
    );

    const buttonEl = getByText("myBtn");
    fireEvent.click(buttonEl);

    expect(refDOM.current).toMatchSnapshot();
  });

  it("Should handle invalid ref", () => {
    const refDOM = {};
    const { getByText } = render(
      <ScrollTo>
        {({ scrollTo }) => (
          <>
            <button
              type="button"
              onClick={() => scrollTo({ ref: refDOM, x: 100, y: 200 })}
            >
              myBtn
            </button>

            <div ref={() => {}}>test</div>
          </>
        )}
      </ScrollTo>
    );

    const buttonEl = getByText("myBtn");
    fireEvent.click(buttonEl);

    expect(window.scrollTo).toHaveBeenCalledTimes(0);
    expect(window.scrollTo.mock.calls[0]).toEqual(undefined);
  });
});
