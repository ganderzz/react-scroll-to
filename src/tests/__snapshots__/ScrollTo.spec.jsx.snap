// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test render prop. Should not break if scrolling by an unknown ID 1`] = `
Object {
  "id": "foo",
  "scrollLeft": 0,
  "scrollTop": 0,
}
`;

exports[`Test render prop. Should render nothing on undefined. 1`] = `<div />`;

exports[`Test render prop. Should render the functional children. 1`] = `
<div>
  <div>
    test
  </div>
</div>
`;

exports[`Test render prop. Should scroll by ID 1`] = `
Object {
  "id": "foo",
  "scrollLeft": 100,
  "scrollTop": 200,
}
`;

exports[`Test render prop. Should scroll by ref when provided 1`] = `
<div>
  test
</div>
`;

exports[`Test render prop. Should update scroll position of ScrollArea's if present, rather than window 1`] = `
Object {
  "scrollLeft": 100,
  "scrollTop": 200,
}
`;
